include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,168 @@ module BYTES-SIMPLIFICATION [symbolic]
+      requires 0 <=Int S1 andBool 0 <=Int W2 andBool S1 +Int lengthBytes(B2) <=Int S2 [simplification]
 
-    rule (#range(#range(BA:Bytes, S0, L0), S1, L1)) => #range(BA, S0 +Int S1, minInt(L1, L0 -Int S1)) +Bytes #buf(maxInt(0, L1 -Int (L0 -Int S1)), 0)
-      requires 0 <=Int S0 andBool 0 <=Int L0 andBool 0 <=Int S1 andBool 0 <=Int L1 [simplification]
+    rule [range-buf-zero]:
+      #range (#buf(W1:Int, X:Int), S2:Int, W2:Int) => #buf(W2, 0)
+      requires 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int (S2 +Int W2))) andBool S2 +Int W2 <=Int W1
+      [simplification]
 
-    // lookup range in memory
-    rule #range(MEM, 0, W) => MEM  requires W ==Int lengthBytes(MEM) [simplification]
+    rule [range-padRightToWidth]:
+      #range(#padRightToWidth(_, BUF), 0, WIDTH) => BUF
+      requires lengthBytes(BUF) ==Int WIDTH [simplification]
 
-    // lookup range in concatenated memory buffers
-    rule #range((BUF1 +Bytes  BUF2), START, WIDTH) =>  #range(BUF2, START -Int lengthBytes(BUF1), WIDTH                        )                                                                    requires                        lengthBytes(BUF1) <=Int START                                                          [simplification]
-    rule #range((BUF1 +Bytes  _   ), START, WIDTH) =>  #range(BUF1, START                       , WIDTH                        )                                                                    requires START +Int WIDTH <=Int lengthBytes(BUF1)                       andBool 0 <=Int START andBool 0 <=Int WIDTH    [simplification]
-    rule #range((BUF1 +Bytes  BUF2), START, WIDTH) => (#range(BUF1, START                       , lengthBytes(BUF1) -Int START)) +Bytes (#range(BUF2, 0, START +Int WIDTH -Int lengthBytes(BUF1)))  requires                        lengthBytes(BUF1) <Int START +Int WIDTH andBool #range(0 <= START < lengthBytes(BUF1)) [simplification]
+    // Memory update
 
-    rule #range(#padRightToWidth(_, BUF), 0, WIDTH) => BUF requires lengthBytes(BUF) ==Int WIDTH [simplification]
+    rule [memUpdate-is-empty]:
+      _:Bytes [ S:Int := _:Bytes ] => .Bytes
+      requires S <Int 0
+      [simplification]
 
-    // WS [ START := WS' ]
+    rule [memUpdate-as-concat-inside]:
+      B1:Bytes [ S:Int := B2:Bytes ] =>
+        #range(B1, 0, S) +Bytes B2 +Bytes #range (B1 , S +Int lengthBytes(B2), lengthBytes(B1) -Int (S +Int lengthBytes(B2)))
+      requires 0 <=Int S andBool S +Int lengthBytes(B2) <Int lengthBytes(B1)
+      [simplification(60)]
 
-    // re-order memory write
-    rule MEM [ K1 := BUF1:Bytes ] [ K2 := BUF2:Bytes ] => MEM [ K2 := BUF2 ] [ K1 := BUF1 ]
-      requires lengthBytes(BUF2) <=Int K1 -Int K2 andBool K2  <Int K1 [simplification]
+    rule [memUpdate-as-concat-outside-1]:
+      B1:Bytes [ S:Int := B2:Bytes ] => #range(B1, 0, S) +Bytes B2
+      requires 0 <=Int S andBool lengthBytes(B1) <=Int S +Int lengthBytes(B2)
+      [simplification(60)]
```

Comment:
Why are we restricting here `lengthBytes(B1) <=Int S +Int lengthBytes(B2)`? Would it be enough to say `lengthBytes(B1) <=Int S` on its own? Also, if we have `lengthBytes(B1) <=Int S`, then we already are guaranteed that `0 <=Int S` because `0 <= Int lengthBytes(_)`.

Reply:
What we have now is more expressive, isn't it? `lengthBytes(B1)` could be less than or equal to `S`, but also in between `S` and `S +Int lengthBytes(B2)`?